HOSTEDCP-2073: Review NodePool API validations and doc

enxebre opened the pull request:
<!--
- Please ensure code changes are split into a series of logically independent commits.
- Every commit should have a subject/title (What) and a description/body (Why).
- Every PR must have a description.
- As an example you can use git commit -m"What" -m"Why" to achieve the requirements above. GitHub automatically recognises the commit description (-m"Why") in single commit PRs and adds it as the PR description.
- Use the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood) in the subject line for every commit. E.g `Mark infraID as required` instead of `This patch marks infraID as required` (This follows Git’s own built-in conventions). See https://github.com/openshift/hypershift/pull/485 as an example.
- See https://hypershift-docs.netlify.app/contribute for more details.

Delete this text before submitting the PR.
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

bryan-cox commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -82,41 +82,47 @@ type NodePool struct {
-	// TODO(dan): Should this be a LocalObjectReference?
-	//
+	// clusterName is the name of the HostedCluster this NodePool belongs to.
+	// If HostedCluster with this name doesn't exist, the controller will no-op until it exists.
```
nit
```suggestion
	// If a HostedCluster with this name doesn't exist, the controller will no-op until it exists.
```

bryan-cox commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -274,8 +273,9 @@ const (
-	//
+	// strategy is the node replacement strategy for nodes in the pool.
+	// In can be either "RollingUpdate" or "OnDelete". RollingUpdate will rollout Nodes honoring maxSurge and maxUnavailable.
+	// OnDelete provide more granular controll and will replace nodes as the old ones are manually deleted.
```
```suggestion
	// OnDelete provide more granular control and will replace nodes as the old ones are manually deleted.
```

bryan-cox commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -359,26 +359,33 @@ type InPlaceUpgrade struct {
+	// upgradeType specifies the type of strategy for handling upgrades.
+	// This can be either "Replace" or "InPlace".
+	// "Replace" will update Nodes by recreating the underlying instances.
+	// "InPlace" will update Nodes by applying changes to the existing instances.
```
This causes the nodes to reboot right? Should we also mention that here?

enxebre replied:
>This causes the nodes to reboot right

Not necessarily, I'll mention it.

bryan-cox reviewed (commented):
Just a few nits and a question.

bryan-cox commented:
/lgtm
/hold
Holding to give others a chance to comment. Feel free to remove the hold when you want.

enxebre commented:
/retest

enxebre commented:
/hold cancel

JoelSpeed commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -479,18 +486,21 @@ type NodePoolCondition struct {
 type Taint struct {
-	// Required. The taint key to be applied to a node.
+	// +required
+	// key is the taint key to be applied to a node.
```
How are taint keys validated?

enxebre replied:
Clarified and added regex

JoelSpeed commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -479,18 +486,21 @@ type NodePoolCondition struct {
-	// The taint value corresponding to the taint key.
+	// value is the taint value corresponding to the taint key.
 	// +optional
 	// +kubebuilder:validation:Pattern:=`^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$`
```
Prefer to use CEL so that you can add a custom message to explain the validation/regex

enxebre commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -82,41 +82,47 @@ type NodePool struct {
+	// release specifies the OCP release used for the NodePool. This informs the
+	// ignition configuration for machines which includes the kubelet version, as well as other platform specific
 	// machine properties (e.g. an AMI on the AWS platform).
+	// It's not supported to use a release in a NodePool which minor version skew against the Control Plane release is bigger than N-2.
```
clarified now.

enxebre commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -135,42 +141,35 @@ type NodePoolSpec struct {
-	// NodeDrainTimeout is the maximum amount of time that the controller will spend on draining a node.
+	// nodeDrainTimeout is the maximum amount of time that the controller will spend on draining a node.
 	// The default value is 0, meaning that the node can be drained without any time limitations.
 	// NOTE: NodeDrainTimeout is different from `kubectl drain --timeout`
```
clarified

enxebre commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -135,42 +141,35 @@ type NodePoolSpec struct {
 	Taints []Taint `json:"taints,omitempty"`
 
-	// PausedUntil is a field that can be used to pause reconciliation on a resource.
+	// pausedUntil is a field that can be used to pause reconciliation on the NodePool controller. Resulting in any change to the NodePool being ignored.
 	// Either a date can be provided in RFC3339 format or a boolean. If a date is
 	// provided: reconciliation is paused on the resource until that date. If the boolean true is
 	// provided: reconciliation is paused on the resource until the field is removed.
 	// +optional
 	PausedUntil *string `json:"pausedUntil,omitempty"`
```
This API doesn't allow breaking changes. I added regex validation.

enxebre commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -184,8 +183,8 @@ type NodePoolSpec struct {
-	// Arch is the preferred processor architecture for the NodePool (currently only supported on AWS)
-	// NOTE: This is set as optional to prevent validation from failing due to a limitation on client side validation with open API machinery:
+	// arch is the preferred processor architecture for the NodePool. Different platforms might have different supported architectures.
+	// TODO: This is set as optional to prevent validation from failing due to a limitation on client side validation with open API machinery:
```
This API doesn't allow breaking changes.

enxebre commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -359,26 +359,33 @@ type InPlaceUpgrade struct {
+	// AutoRepair will no-op when more than 2 Nodes are unhealthy at the same time. Giving time for the cluster to stabilize or to the user to manually intervene.
 	// +optional
 	// +kubebuilder:default=false
 	AutoRepair bool `json:"autoRepair"`
```
This API doesn't allow breaking changes.

enxebre commented:
/hold
Addressed all feedback but Image, pausedUntil and taint regex are failing when applying the CRD. I'll fix those.

enxebre commented:
>Addressed all feedback but Image, pausedUntil and taint regex are failing when applying the CRD. I'll fix those.

Fixed escaping for Image, pausedUntil and used // +kubebuilder:validation:Pattern for taints key/value with a TODO since // +kubebuilder:validation:XValidation errors exceeding budget

bryan-cox commented on api/hypershift/v1beta1/nodepool_types.go:
```diff
@@ -82,95 +82,98 @@ type NodePool struct {
 	// Either a date can be provided in RFC3339 format or a boolean. If a date is
 	// provided: reconciliation is paused on the resource until that date. If the boolean true is
 	// provided: reconciliation is paused on the resource until the field is removed.
+	// +kubebuilder:validation:MaxLength=100
```
nit: wouldn't the max length be something like 35, for a RFC3339 date example?

bryan-cox reviewed (commented):
Just one nit comment

enxebre commented:
Added a bunch of e2e on create UX tests and split commits into api changes/tests and make update

enxebre commented:
>Fixed escaping for Image, pausedUntil and used // +kubebuilder:validation:Pattern for taints key/value with a TODO since // +kubebuilder:validation:XValidation errors exceeding budget

Fixed the budget errors by adding maxItems to taints

enxebre commented:
/test security

enxebre commented:
/hold cancel

bryan-cox commented:
/lgtm
